matrix_openCL/Project1/main.cpp

```diff
@@ -0,0 +1,88 @@
+#define __CL_ENABLE_EXCEPTIONS
+#include <CL/cl.hpp>
+#include <utility>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+#include <array>
+using namespace std;
+using namespace cl;
+
+int main() {
+    // Create the two input vectors
+    const int LIST_SIZE = 1000;
+    array<int, LIST_SIZE> A = {};
+    array<int, LIST_SIZE> B = {};
+    for (int i = 0; i < LIST_SIZE; i++) {
+        A[i] = i;
+        B[i] = LIST_SIZE - i;
+    }
+
+    try {
+        // Get available platforms
+        vector<Platform> platforms;
+        Platform::get(&platforms);
+
+        // Select the default platform and create a context using this platform and the GPU
+        cl_context_properties cps[3] = {
+            CL_CONTEXT_PLATFORM,
+            (cl_context_properties)(platforms[0])(),
+            0
+        };
+        Context context(CL_DEVICE_TYPE_GPU, cps);
+
+        // Get a list of devices on this platform
+        vector<Device> devices = context.getInfo<CL_CONTEXT_DEVICES>();
+
+        // Create a command queue and use the first device
+        CommandQueue queue = CommandQueue(context, devices[0]);
+
+        // Read source file
+        std::ifstream sourceFile("vector_adds_kernel.cl");
+        std::string sourceCode(
+            std::istreambuf_iterator<char>(sourceFile),
+            (std::istreambuf_iterator<char>()));
+        Program::Sources source(1, std::make_pair(sourceCode.c_str(), sourceCode.length() + 1));
+
+        // Make program of the source code in the context
+        Program program = Program(context, source);
+
+        // Build program for these specific devices
+        program.build(devices);
+
+        // Make kernel
+        Kernel kernel(program, "vector_add");
+
+        // Create memory buffers
+        Buffer bufferA = Buffer(context, CL_MEM_READ_ONLY, LIST_SIZE * sizeof(int));
+        Buffer bufferB = Buffer(context, CL_MEM_READ_ONLY, LIST_SIZE * sizeof(int));
+        Buffer bufferC = Buffer(context, CL_MEM_WRITE_ONLY, LIST_SIZE * sizeof(int));
+
+        // Copy lists A and B to the memory buffers
+        queue.enqueueWriteBuffer(bufferA, CL_TRUE, 0, LIST_SIZE * sizeof(int), &A[0]);
+        queue.enqueueWriteBuffer(bufferB, CL_TRUE, 0, LIST_SIZE * sizeof(int), &B[0]);
+
+        // Set arguments to kernel
+        kernel.setArg(0, bufferA);
+        kernel.setArg(1, bufferB);
+        kernel.setArg(2, bufferC);
+
+        // Run the kernel on specific ND range
+        NDRange global(LIST_SIZE);
+        NDRange local(1);
+        queue.enqueueNDRangeKernel(kernel, NullRange, global, local);
```

Comment:
I don't fully understand how this function works. From what i see it executes kernel global times. From documentation i see that NullRange is an offset. Offset of what? What does local do?

Reply:
again, a quick google search tells me that this function executes the kernel, global is the number of workitems to spawn. each work item executes the function vector_add. in this case, since `vector_ad` works on just one item in the vector, we need to create LIST_SIZE work items, one for each element in the vector.
local is the number of work items per work groups. we don't need to make it more than 1 here
the NullRange offset is for calculating the global id of the work item. all work items start from zero up to the number of work items that we specify (in our case the variable `global=LIST_SIZE`). if we set an offset the IDs will start form that offset instead of zero. the work items that are executing our vector_add function get the id here: `int i = get_global_id(0);`.

You can for example use the offset to calculate the sum for just the second half of the vector. for that you define global as `LIST_SIZE/2` and the offset also as `LIST_SIZE/2`:
```
queue.enqueueNDRangeKernel(kernel, NDRange(LIST_SIZE/2), NDRange(LIST_SIZE/2), local);
```
This will create just LIST_SIZE/2 work items, and the ID of each work item (which is used as the index in the vector) will start at LIST_SIZE/2. thus calculating the sum for just the second half of the vector.